### PR TITLE
Entity linking eng

### PR DESCRIPTION
**Pull request by dmitrijeuseew**
Config for entity linking in English

### CODE REVIEW

File: deeppavlov/models/kbqa/kbqa_entity_linking.py [hunk 1]
```diff
@@ -171,20 +201,35 @@ def __call__(self, entity_substr_batch: List[List[str]], entity_positions_batch:
 
         return entity_ids_batch, confidences_batch
 
-    def link_entity(self, entity: str, context: str = None) -> Tuple[List[str], List[float]]:
+    def link_entity(self, entity: str, context: str = None, template_found: str = None, cut_entity: bool = False) -> \
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/kbqa_entity_linking.py:204]: `variable: type = None` -> `varialbe: Optional[type] = None`

File: deeppavlov/models/kbqa/query_generator.py [hunk 2]
```diff
@@ -62,15 +63,18 @@ def __init__(self, wiki_parser: WikiParser,
...
                  template_type_batch: Union[List[List[str]], List[str]],
                  entities_from_ner_batch: List[List[str]],
                  types_from_ner_batch: List[List[str]]) -> List[Union[List[Tuple[str, Any]], List[str]]]:
 
         candidate_outputs_batch = []
-        for question, template_type, entities_from_ner, types_from_ner in \
-                zip(question_batch, template_type_batch, entities_from_ner_batch, types_from_ner_batch):
+        for question, question_sanitized, template_type, entities_from_ner, types_from_ner in \
+                zip(question_batch, question_san_batch, template_type_batch,
+                                                        entities_from_ner_batch, types_from_ner_batch):
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/query_generator.py:74]:
```suggestion
                    entities_from_ner_batch, types_from_ner_batch):
```

File: deeppavlov/models/kbqa/query_generator.py [hunk 3]
```diff
@@ -147,14 +154,16 @@ def query_parser(self, question: str, query_info: Dict[str, str],
         entity_combs = make_combs(selected_entity_ids, permut=True)
         type_combs = make_combs(selected_type_ids, permut=False)
         log.debug(f"(query_parser)entity_combs: {entity_combs[:3]}, type_combs: {type_combs[:3]}, rel_combs: {rel_combs[:3]}")
+        confidence = 0.0
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/query_generator.py:157]: redundant declaration: you redeclare this variable two lines below

File: deeppavlov/models/kbqa/query_generator.py [hunk 4]
```diff
@@ -147,14 +154,16 @@ def query_parser(self, question: str, query_info: Dict[str, str],
         entity_combs = make_combs(selected_entity_ids, permut=True)
         type_combs = make_combs(selected_type_ids, permut=False)
         log.debug(f"(query_parser)entity_combs: {entity_combs[:3]}, type_combs: {type_combs[:3]}, rel_combs: {rel_combs[:3]}")
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/query_generator.py:156]:
```suggestion
        log.debug(f"(query_parser)entity_combs: {entity_combs[:3]}, type_combs: {type_combs[:3]},"
                  f" rel_combs: {rel_combs[:3]}")
```
too long line

File: deeppavlov/models/kbqa/query_generator.py [hunk 5]
```diff
@@ -147,14 +154,16 @@ def query_parser(self, question: str, query_info: Dict[str, str],
...
         for comb_num, combs in enumerate(itertools.product(entity_combs, type_combs, rel_combs)):
+            confidence = np.prod([score for rel, score in combs[2][:-1]])
             query_hdt_seq = [
                 fill_query(query_hdt_elem, combs[0], combs[1], combs[2]) for query_hdt_elem in query_sequence]
             if comb_num == 0:
                 log.debug(f"\n_______________________________\nfilled query: {query_hdt_seq}\n_______________________________\n")
             candidate_output = self.wiki_parser(
                 rels_from_query + answer_ent, query_hdt_seq, filter_info, order_info)
-            candidate_outputs += [combs[2][:-1] + output for output in candidate_output]
+            candidate_outputs += [[rel for rel, score in combs[2][:-1]] + output + [confidence] for output in candidate_output]
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/query_generator.py:166]:
```suggestion
            candidate_outputs += [[rel for rel, score in combs[2][:-1]] + output + [confidence]
                                  for output in candidate_output]
```
too long line

File: deeppavlov/models/kbqa/query_generator_online.py [hunk 6]
```diff
@@ -63,15 +64,18 @@ def __init__(self, wiki_parser: WikiParserOnline,
...
                  template_type_batch: List[str],
                  entities_from_ner_batch: List[List[str]],
                  types_from_ner_batch: List[List[str]]) -> List[Union[List[Tuple[str, Any]], List[str]]]:
 
         candidate_outputs_batch = []
-        for question, template_type, entities_from_ner, types_from_ner in \
-                zip(question_batch, template_type_batch, entities_from_ner_batch, types_from_ner_batch):
+        for question, question_sanitized, template_type, entities_from_ner, types_from_ner in \
+                zip(question_batch, question_san_batch, template_type_batch,
+                                                        entities_from_ner_batch, types_from_ner_batch):
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/query_generator_online.py:75]:
```suggestion
                zip(question_batch, question_san_batch, template_type_batch,
                    entities_from_ner_batch, types_from_ner_batch):
```

File: deeppavlov/models/kbqa/query_generator_online.py [hunk 7]
```diff
@@ -63,15 +64,18 @@ def __init__(self, wiki_parser: WikiParserOnline,
...
 
-            candidate_outputs = self.find_candidate_answers(question, template_type, entities_from_ner, types_from_ner)
+            candidate_outputs = self.find_candidate_answers(question, question_sanitized,
+                                                            template_type, entities_from_ner, types_from_ner)
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/query_generator_online.py:78]: teplate_type: Expected type 'List[str]', got 'str' instead

File: deeppavlov/models/kbqa/query_generator_online.py [hunk 8]
```diff
@@ -157,16 +161,20 @@ def query_parser(self, question: str, query_info: Dict[str, str],
...
+                log.debug(f"\n_______________________________\nfilled query: {filled_query}\n_______________________________\n")
             candidate_output = self.wiki_parser.get_answer(filled_query)
             
             out_vars = filter_rels + rels_from_query + answer_ent
             candidate_output = [output for output in candidate_output
                 if (all([filter_value in output[filter_var[1:]]["value"] for filter_var, filter_value in property_types.items()])
                 and all([not output[ent[1:]]["value"].startswith("http://www.wikidata.org/value") for ent in answer_ent]))]
-            candidate_outputs += [combs[2][:-1] + [output[var[1:]]["value"] for var in out_vars] for output in candidate_output]
+            candidate_outputs += [combs[2][:-1] + [output[var[1:]]["value"] for var in out_vars] + [confidence]
+                                                                                    for output in candidate_output]
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/query_generator_online.py:177]:
```suggestion
                                  for output in candidate_output]
```

File: deeppavlov/models/kbqa/tree_to_sparql.py [hunk 9]
```diff
@@ -29,21 +34,129 @@
 log = getLogger(__name__)
 
 
+@register('adj_to_noun')
+class AdjToNoun:
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/tree_to_sparql.py:38]: add to `docs/apiref/models/kbqa.rst`

File: deeppavlov/models/kbqa/tree_to_sparql.py [hunk 10]
```diff
@@ -29,21 +34,129 @@
...
+        indices = np.array(indices)
+        shape=(len(words), self.max_word_length*self.alphabet_length)
+
+        matrix = csr_matrix((data, indices, indptr), shape=(len(words), self.max_word_length*self.alphabet_length))
+
+        return matrix
+
+
+@register('udpipe_parser')
+class UdpipeParser(Component):
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/tree_to_sparql.py:116]: add to `docs/apiref/models/kbqa.rst`

**Review by IgnatovFedor — COMMENTED** [review]
Add entity_linking_eng to test_quick_start, add description to documentation

File: deeppavlov/models/kbqa/tree_to_sparql.py [hunk 11]
```diff
@@ -148,14 +287,17 @@ def find_clause_node(self, root: Node, unknown_branch: Node) -> Tuple[Node]:
...
             else:
                 used_desc = node.children
             for elem in used_desc:
                 if (not cut_clause or (cut_clause and elem.deprel != "acl")) and elem not in conj_list \
                     and (elem.deprel != "appos" or (elem.deprel == "appos" \
-                    and (not elem.children or (len(elem.children) == 1 and elem.children[0].deprel == "flat:name")))):
+                    and (not elem.children or (len(elem.children) == 1 and elem.children[0].deprel == "flat:name")
+                    or (len(elem.children) > 1 and ("«" in [nd.form for nd in elem.children]
+                    or '"' in [nd.form for nd in elem.children]
+                    or '``' in [nd.form for nd in elem.children]))))):
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/tree_to_sparql.py:300]: This is quite frightening. Could you decompose this condition into multiple `if`s or simplify it in other way? For example, `("«" in [nd.form for nd in elem.children] or '"' in [nd.form for nd in elem.children] or '``' in [nd.form for nd in elem.children])` could replaced with `{"«", '"', '``'} & {nd.form for nd in elem.children}`

File: deeppavlov/models/kbqa/tree_to_sparql.py [hunk 12]
```diff
@@ -218,7 +361,11 @@ def find_year_or_number(self, node: Node) -> bool:
 
     def find_appos_tokens(self, node: Node, appos_token_nums: List[int]) -> List[int]:
         for elem in node.children:
-            if elem.deprel == "appos" and len(elem.descendants) > 1 or (len(elem.descendants) == 1 and elem.descendants[0].deprel != "flat:name"):
+            if elem.deprel == "appos" and (len(elem.descendants) > 1 and
+                    "«" not in [nd.form for nd in elem.children] and
+                    '"' not in [nd.form for nd in elem.children] and
+                    '``' not in [nd.form for nd in elem.children] or
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/tree_to_sparql.py:367]: same as comment on lines +297 to +300

File: deeppavlov/models/kbqa/tree_to_sparql.py [hunk 13]
```diff
@@ -52,16 +165,20 @@ def __call__(self, syntax_tree_batch: List[str],
...
+            if root:
+                log.debug(f"syntax tree info, root: {root.form}")
+                unknown_node, unknown_branch = self.find_branch_with_unknown(root)
             positions = [num for position in positions for num in position]
             if unknown_node:
                 log.debug(f"syntax tree info, unknown node: {unknown_node.form}, unknown branch: {unknown_branch.form}")
+                log.debug(f"wh_leaf: {self.wh_leaf}")
                 clause_node, clause_branch = self.find_clause_node(root, unknown_branch)
                 modifiers, clause_modifiers = self.find_modifiers_of_unknown(unknown_node)
                 log.debug(f"modifiers: {[modifier.form for modifier in modifiers]}")
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/tree_to_sparql.py:184]: comment on `["каким был", "какой была"]` from line 185: what if language is English?

File: deeppavlov/models/kbqa/tree_to_sparql.py [hunk 14]
```diff
@@ -71,68 +188,90 @@ def __call__(self, syntax_tree_batch: List[str],
...
+                    self.root_entity = True
                 query_nums, entities_dict, types_dict = self.build_query(new_root, unknown_branch, root_desc,
-                                                                         unknown_node, modifiers, clause_modifiers, positions)
-
-                question = ' '.join([node.form for node in tree.descendants if (node.ord not in appos_token_nums or node.ord not in clause_token_nums)])
+                                     unknown_node, modifiers, clause_modifiers, clause_node, positions, count=count)
+    
+                if self.lang == "rus":
+                    question = ' '.join([node.form for node in tree.descendants \
+                        if (node.ord not in appos_token_nums or node.ord not in clause_token_nums)])
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/tree_to_sparql.py:214]: `clause_token_nums` haven't declared

File: deeppavlov/models/kbqa/tree_to_sparql.py [hunk 15]
```diff
@@ -236,8 +383,24 @@ def find_clause_tokens(self, node: Node, clause_node: Node, clause_token_nums: L
...
+        first_or_last = ""
+        nodes = [node]
+        while nodes:
+            for node in nodes:
+                node_desc = defaultdict(set)
+                for elem in node.children:
+                    node_desc[elem.deprel].add(self.morph.parse(elem.form.lower())[0].inflect({"masc", "sing", "nomn"}).word)
+                if "amod" in node_desc.keys() and "nmod" in node_desc.keys() and \
+                   node_desc["amod"].intersection({"первый", "последний"}):
+                    first_or_last = node_desc["amod"].intersection({"первый", "последний"})
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/tree_to_sparql.py:396]: what if language is English?

**Review by IgnatovFedor — CHANGES_REQUESTED** [review]

File: deeppavlov/models/kbqa/query_generator_online.py [hunk 16]
```diff
@@ -157,16 +161,20 @@ def query_parser(self, question: str, query_info: Dict[str, str],
         log.debug(f"(query_parser)entity_combs: {entity_combs[:3]}")
         type_combs = make_combs(selected_type_ids, permut=False)
         log.debug(f"(query_parser)type_combs: {type_combs[:3]}")
-        for combs in itertools.product(entity_combs, type_combs, rel_combs):
+        confidence = 0.0
+        for comb_num, combs in enumerate(itertools.product(entity_combs, type_combs, rel_combs)):
             filled_query, filter_rels = fill_online_query(query, combs[0], combs[1], combs[2], fill_rel_variables,
                                                           filter_rel_variables, rels_list_for_filter)
+            if comb_num == 0:
+                log.debug(f"\n_______________________________\nfilled query: {filled_query}\n_______________________________\n")
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/models/kbqa/query_generator_online.py:169]: I inserted it to make filled query more visible when debugging.

File: deeppavlov/configs/kbqa/kbqa_cq_rus.json [hunk 17]
```diff
@@ -29,9 +29,16 @@
         "in": ["x"],
         "out": ["syntax_info"]
       },
+      {
+        "class_name": "adj_to_noun",
+        "freq_dict_filename": "{DOWNLOADS_PATH}/wikidata_rus/freqrnc2011.csv",
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/configs/kbqa/kbqa_cq_rus.json:34]: Loaded to share

**Review by IgnatovFedor — APPROVED** [review]

**Pull request merged**